### ISSUE TITLE
BigPanda Integration

### ISSUE THREAD

**Issue by y-ard** [influxdata/kapacitor#2304]
Hello!

It'll be nice to have a preconfigured integration  with BigPanda, which will simplify upstream alert integration. This is simple HTTPS RESTful API, and to have it out of the box will be great.

Thanks,

~yuri

**Comment by russorat** [comment]
@y-ard thanks! Have you tried configuring the Http POST endpoint to talk to BigPanda? If so, do you mind sharing the config?

**Comment by y-ard** [comment]
Not yet. We still waiting on the environment 😊 nothing is flying this days 😊

**Comment by y-ard** [comment]
Hi Russ,

Finally I got the test environment and I did. So far it doesn’t work ☹.

Yes can provide you with our configurations, and the way how is alert should be configured

Thanks,

~yuri

**Comment by y-ard** [comment]
Well so far no success. At least major one :(
even if i am sending legit json BigPanda responds with:
err="<!DOCTYPE html>\n<html lang=\"en\">\n<head>\n<meta charset=\"utf-8\">\n<title>Error</title>\n</head>\n<body>\n<pre>Bad Request</pre>\n</body>\n</html>\n" code=400

if the same body submitted with curl command it makes it through.

plus we have some set of required fields, like "host", "cluster", etc. But i did not figured out yet how to populate it, with "alert-template".

**Comment by timhallinflux** [comment]
While the event handler has been created and it represents a good first pass.  It does not currently expose the mechanism to populate the `tags` parameter. https://docs.bigpanda.io/reference#alert-object

The is a critical part of the contextual information which needs to be passed upstream to BigPanda.  

Need to extend the existing integration to allow for this to be passed.  Further, examples should be documented about how to construct the various key value pairs within a TICKscript to build up the appropriate context.

**Comment by rhajek** [comment]
Alerts Object with the `tags` attribute is related to the `GET` method https://api.bigpanda.io/resources/v1.0/incidents/{incident ID}/alerts/ used to read the schema of the alert incident.

We use `POST` method https://api.bigpanda.io/data/v2/alerts for posting alerts.

BigPanda handler automatically adds all tags presented in `alert.EventData` .
Note: If the tick script contains `groupBy` operator, then resulting alert event contains only tags specified in by the `groupBy` expression.

#### Example for global tags:

telegraf global tags config:

```
[global_tags]
   dc = "us-east-1"
   rack = "1a"
   user = "$USER"%        
```

Tick Script
```
var data = stream
  |from()
    .database('telegraf')
    .retentionPolicy('autogen')
    .measurement('cpu')
    .groupBy('host', 'dc', 'rack', 'user')
    .where(lambda: "cpu" == 'cpu-total')
  |eval(lambda: 100.0 - "usage_idle")
    .as('total_used')
  |window()
    .period(10s)
    .every(10s)
  |mean('total_used')
    .as('total_used')
  |alert()
    .id('cpu_usage_check')
    .message('Hey {{ index .Tags "host"}} / {{ .ID }}: is high!')
    .details('https://example.com/dashboard/{{ index .Tags "host"}}')
    .info(lambda: "total_used" > 70)
    .warn(lambda: "total_used" > 80)
    .crit(lambda: "total_used" > 90)
    .stateChangesOnly()
    .bigPanda()
```

Generated BigPanda alert contains all tags from `groupBy` expression:

```json
{
  "app_key": "8f5795cada387b9f310f41fc369c88e4",
  "check": "cpu_usage_check",
  "dc": "us-east-1",
  "description": "Hey kozel.local / cpu_usage_check: is high!",
  "details": "https://example.com/dashboard/kozel.local",
  "host": "kozel.local",
  "rack": "1a",
  "status": "critical",
  "task": "cpu_usage_check:cpu",
  "timestamp": 1633010510,
  "total_used": "92.53376688320374",
  "user": "rob"
}
```

If you need additional custom tag `your_tag` into alert message, you can use `eval` expression like this:

```
var data = stream
  |from()
    .database('telegraf')
    .retentionPolicy('autogen')
    .measurement('cpu')
    .groupBy('host', 'dc', 'rack', 'user')
    .where(lambda: "cpu" == 'cpu-total')
  |eval(lambda: 100.0 - "usage_idle")
    .as('total_used')
  |window()
    .period(10s)
    .every(10s)
  |mean('total_used')
    .as('total_used')
  |eval(lambda: 'your_tag_value')
                  .as('your_tag')
                  .tags('your_tag')
                  .keep()
  |alert()
    .id('cpu_usage_check')
    .message('Hey {{ index .Tags "host"}} / {{ .ID }}: is high!')
    .details('https://example.com/dashboard/{{ index .Tags "host"}}')
    .info(lambda: "total_used" > 70)
    .warn(lambda: "total_used" > 80)
    .crit(lambda: "total_used" > 90)
    .stateChangesOnly()
    .bigPanda()
```

Then the resulting BigPanda payload will 

```json
{
  "app_key": "8f5795cada387b9f310f41fc369c88e4",
  "check": "cpu_usage_check",
  "dc": "us-east-1",
  "description": "Hey kozel.local / cpu_usage_check: is high!",
  "details": "https://example.com/dashboard/kozel.local",
  "host": "kozel.local",
  "rack": "1a",
  "status": "critical",
  "task": "cpu_usage_check:cpu",
  "timestamp": 1633011490,
  "total_used": "100",
  "user": "rob",
  "your_tag": "your_tag_value"
}
```

I think that there is not need to modify existing BigPanda handler implementation.

**Comment by y-ard** [comment]
Hi Robert,

Unfortunately but if you’ll take a look into our example, you probably will see, that we need to have an additional fields which we are using: Grafana_link, runbook_link, source, and some others.
And all of them unfortunately will be left uncovered with your current approach, and make it unusable for us.

Thanks,

~yuri

From: Robert Hajek ***@***.***>
Reply to: influxdata/kapacitor ***@***.***>
Date: Thursday, September 30, 2021 at 7:29 AM
To: influxdata/kapacitor ***@***.***>
Cc: Yuri Ardulov ***@***.***>, Mention ***@***.***>
Subject: [External Sender] Re: [influxdata/kapacitor] BigPanda Integration (#2304)


Alerts Object with the tags attribute is related to the GET method https://api.bigpanda.io/resources/v1.0/incidents/{incident<https://urldefense.proofpoint.com/v2/url?u=https-3A__api.bigpanda.io_resources_v1.0_incidents_-257Bincident&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=AtThdJAj8wo0UUHeu49aw37mEqTJkFbS8SrFs8zxdgY&s=bmTQFJA4Sm4ukH5XPsXpCVHa4Z3Fxio0-ICJrduHrss&e=> ID}/alerts/ used to read the schema of the alert incident.

We use POST method https://api.bigpanda.io/data/v2/alerts<https://urldefense.proofpoint.com/v2/url?u=https-3A__api.bigpanda.io_data_v2_alerts&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=AtThdJAj8wo0UUHeu49aw37mEqTJkFbS8SrFs8zxdgY&s=Z-Gp_WkTqQMhDmuze8apWySs3jfYiG56sMl4xe00GJg&e=> for posting alerts.

BigPanda handler automatically adds all tags presented in alert.EventData .
Note: If the tick script contains groupBy operator, then resulting alert event contains only tags specified in by the groupBy expression.

Example for global tags:

telegraf global tags config:

[global_tags]

   dc = "us-east-1"

   rack = "1a"

   user = "$USER"%

Tick Script

var data = stream

  |from()

    .database('telegraf')

    .retentionPolicy('autogen')

    .measurement('cpu')

    .groupBy('host', 'dc', 'rack', 'user')

    .where(lambda: "cpu" == 'cpu-total')

  |eval(lambda: 100.0 - "usage_idle")

    .as('total_used')

  |window()

    .period(10s)

    .every(10s)

  |mean('total_used')

    .as('total_used')

  |alert()

    .id('cpu_usage_check')

    .message('Hey {{ index .Tags "host"}} / {{ .ID }}: is high!')

    .details('https://example.com/dashboard/{{ index .Tags "host"}}')

    .info(lambda: "total_used" > 70)

    .warn(lambda: "total_used" > 80)

    .crit(lambda: "total_used" > 90)

    .stateChangesOnly()

    .bigPanda()

Generated BigPanda alert contains all tags from groupBy expression:

{

  "app_key": "8f5795cada387b9f310f41fc369c88e4",

  "check": "cpu_usage_check",

  "dc": "us-east-1",

  "description": "Hey kozel.local / cpu_usage_check: is high!",

  "details": "https://example.com/dashboard/kozel.local",

  "host": "kozel.local",

  "rack": "1a",

  "status": "critical",

  "task": "cpu_usage_check:cpu",

  "timestamp": 1633010510,

  "total_used": "92.53376688320374",

  "user": "rob"

}

If you need additional custom tag your_tag into alert message, you can use eval expression like this:

var data = stream

  |from()

    .database('telegraf')

    .retentionPolicy('autogen')

    .measurement('cpu')

    .groupBy('host', 'dc', 'rack', 'user')

    .where(lambda: "cpu" == 'cpu-total')

  |eval(lambda: 100.0 - "usage_idle")

    .as('total_used')

  |window()

    .period(10s)

    .every(10s)

  |mean('total_used')

    .as('total_used')

  |eval(lambda: 'your_tag_value')

                  .as('your_tag')

                  .tags('your_tag')

                  .keep()

  |alert()

    .id('cpu_usage_check')

    .message('Hey {{ index .Tags "host"}} / {{ .ID }}: is high!')

    .details('https://example.com/dashboard/{{ index .Tags "host"}}')

    .info(lambda: "total_used" > 70)

    .warn(lambda: "total_used" > 80)

    .crit(lambda: "total_used" > 90)

    .stateChangesOnly()

    .bigPanda()

Then the resulting BigPanda payload will

{

  "app_key": "8f5795cada387b9f310f41fc369c88e4",

  "check": "cpu_usage_check",

  "dc": "us-east-1",

  "description": "Hey kozel.local / cpu_usage_check: is high!",

  "details": "https://example.com/dashboard/kozel.local",

  "host": "kozel.local",

  "rack": "1a",

  "status": "critical",

  "task": "cpu_usage_check:cpu",

  "timestamp": 1633011490,

  "total_used": "100",

  "user": "rob",

  "your_tag": "your_tag_value"

}

I think that there is not need to modify existing BigPanda handler implementation.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_influxdata_kapacitor_issues_2304-23issuecomment-2D931376053&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=AtThdJAj8wo0UUHeu49aw37mEqTJkFbS8SrFs8zxdgY&s=lGcxuWqS3Zs7qcWUgzOHv3ahms1K8vwYhAA_529iJrk&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AOX5HSXRO52F7FKCT7ACKZLUERX6BANCNFSM4LCWADIQ&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=AtThdJAj8wo0UUHeu49aw37mEqTJkFbS8SrFs8zxdgY&s=MlnjAW6st9GDiAG46-6FV-OMgr3AyJoNgGfs_m-Mfaw&e=>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=AtThdJAj8wo0UUHeu49aw37mEqTJkFbS8SrFs8zxdgY&s=EoJ1mP88VUyP-hN4WqbqNZZ8O6dwIzjot5tDB4mqyw8&e=> or Android<https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=AtThdJAj8wo0UUHeu49aw37mEqTJkFbS8SrFs8zxdgY&s=Ih6yfQ7S-5D-j5l0VPfE0bAcCd0ek42iIF0jiabpR_g&e=>.

**Comment by alespour** [comment]
@y-ard I'm close to submitting a PR that adds `attribute` option to BigPanda handler. It will allow users to add [Additional alert attributes](https://docs.bigpanda.io/reference#alerts) to the payload.

Given the above example with additional attributes set
```
 ...
  |alert()
    .id('cpu_usage_check')
    .message('Hey {{ index .Tags "host"}} / {{ .ID }}: is high!')
    .details('https://example.com/dashboard/{{ index .Tags "host"}}')
    .info(lambda: "total_used" > 70)
    .warn(lambda: "total_used" > 80)
    .crit(lambda: "total_used" > 90)
    .stateChangesOnly()
    .bigPanda()
        .attribute('a_grafana_link', '{{.Tags.grafana_link}}')
        .attribute('z_full_message', '{{.Tags}}')
        .attribute('sourcehost': 'qa-ml4-grafn02')
```
The output
```
{
  "app_key": "8f5795cada387b9f310f41fc369c88e4",
  "a_grafana_link", "http://example.com/grafana",
  "check": "cpu_usage_check",
  "dc": "us-east-1",
  "description": "Hey kozel.local / cpu_usage_check: is high!",
  "details": "https://example.com/dashboard/kozel.local",
  "host": "kozel.local",
  "rack": "1a",
  "sourcehost": "qa-ml4-grafn02",
  "status": "critical",
  "task": "cpu_usage_check:cpu",
  "timestamp": 1633011490,
  "total_used": "100",
  "user": "rob",
  "z_full_message": "map[host:kozel.local,dc:us-east-1,rack:1a,user:rob]"
}
```

I have 2 questions:
1. would this be acceptable solution for the customer (Yuri?)
2. as of now, the bigpanda handler adds all tags (`host`, `rack`, `dc` and `user` in the example) and fields (`total_used`) as additional attributes - **is this really requested behavior**?

Thank you.

**Comment by y-ard** [comment]
Alespour,

This will be an excellent solution.

Thanks,

~yuri

From: alespour ***@***.***>
Reply to: influxdata/kapacitor ***@***.***>
Date: Thursday, October 14, 2021 at 1:31 AM
To: influxdata/kapacitor ***@***.***>
Cc: Yuri Ardulov ***@***.***>, Mention ***@***.***>
Subject: [External Sender] Re: [influxdata/kapacitor] BigPanda Integration (#2304)


yuri,

the following template data are available for the new attribute value expression:

ID

Message

Level

Time

Duration

Details

Name

TaskName

Group

Tags

Fields

Fields are query result values. With a script like the above, a value from query result can be used as additional attribute value like

  ...

  |from()

    ...

  |eval(...)

    .as('total_used')

  |alert()

    ...

    .bigPanda()

         .attribute('x_value', '{{ .Fields.total_used }}')

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_influxdata_kapacitor_issues_2304-23issuecomment-2D943135502&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=wwt6CcodMzgauWnk5n1-g6keJ1wCpYc6B1JR5KOYOuM&s=EWzoWVcjfeVrnUGvc0f7O3TvGpgzEQzR7BsxP57-was&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AOX5HSUDYMORMCUBJYSAN63UG2IPHANCNFSM4LCWADIQ&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=wwt6CcodMzgauWnk5n1-g6keJ1wCpYc6B1JR5KOYOuM&s=L4SxdssnqanZIUqtoI_Xhdse7PD7lIQRbvHEv9j2DDM&e=>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=wwt6CcodMzgauWnk5n1-g6keJ1wCpYc6B1JR5KOYOuM&s=r3eSHhuFVzSN_i75Oi4X0vCrgxzW_M2SsbQu6yiiUP4&e=> or Android<https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=DS6PUFBBr_KiLo7Sjt3ljp5jaW5k2i9ijVXllEdOozc&r=2yDYLreFh9x0BVtKWxkd6HoBeq3WmPOH5Hgx18SMZ0E&m=wwt6CcodMzgauWnk5n1-g6keJ1wCpYc6B1JR5KOYOuM&s=EtNn3vlC8DFeXsFDnyWd0UFdha6qCxQ9pQFQqgoeVWc&e=>.

**Comment by timhallinflux** [comment]
this has been resolved via #2643
